# DATA WRANGLING

# Amazon reviews and ratings - Amazon Appliances

In [1]:
import pandas as pd
import numpy as np

In [2]:
review_df = pd.read_json('../data/Appliances_5.json', lines=True)

In [3]:
review_df.head(3)

overall  verified   reviewTime      reviewerID        asin  \
0        5      True  08 22, 2013  A34A1UP40713F8  B00009W3I4   
1        5      True   02 8, 2016  A1AHW6I678O6F2  B00009W3PA   
2        5      True   08 5, 2015   A8R48NKTGCJDQ  B00009W3PA   

                       style   reviewerName  \
0  {'Style:': ' Dryer Vent'}  James. Backus   
1       {'Size:': ' 6-Foot'}         kevin.   
2       {'Size:': ' 6-Foot'}      CDBrannom   

                                          reviewText        summary  \
0  I like this as a vent as well as something tha...  Great product   
1                                          good item     Five Stars   
2                     Fit my new LG dryer perfectly.     Five Stars   

   unixReviewTime vote image  
0      1377129600  NaN   NaN  
1      1454889600  NaN   NaN  
2      1438732800  NaN   NaN

In [4]:
review_df.shape

(2277, 12)

In [5]:
review_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2277 entries, 0 to 2276
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   overall         2277 non-null   int64 
 1   verified        2277 non-null   bool  
 2   reviewTime      2277 non-null   object
 3   reviewerID      2277 non-null   object
 4   asin            2277 non-null   object
 5   style           38 non-null     object
 6   reviewerName    2277 non-null   object
 7   reviewText      2277 non-null   object
 8   summary         2277 non-null   object
 9   unixReviewTime  2277 non-null   int64 
 10  vote            2074 non-null   object
 11  image           828 non-null    object
dtypes: bool(1), int64(2), object(9)
memory usage: 198.0+ KB


<b> The features which have non-null values include:
    
    * overall : rating of the product
    * verified 
    * reviewTime : time of the review (raw)
    * reviewerID : ID of the reviewer
    * asin : ID of the product
    * unixReviewTime : time of the review (unix time)
    * reviewerName : name of the reviewer
    * summary : summary of the review
    * reviewerText : text of the review
    
   The features which have considerable amount of null values include:
    
    * vote : helpful votes of the review
    * image : images that users post after they have received the product
    
   The features that have less than 100 null values include:
    
    * style : a dictionary of the product metadata
     
    
   We will need overall, reviewerID, asin, reviewerText further to work on our model. We can select other features like reviewerName, but we already have reviewerID, so this feature is not necessary.

In [6]:
# Renaming column names : asin to productID and overall to rating.
review_df.rename(columns={"asin":"productID","overall":"rating"},inplace=True)
review_df.head()

rating  verified   reviewTime      reviewerID   productID  \
0       5      True  08 22, 2013  A34A1UP40713F8  B00009W3I4   
1       5      True   02 8, 2016  A1AHW6I678O6F2  B00009W3PA   
2       5      True   08 5, 2015   A8R48NKTGCJDQ  B00009W3PA   
3       5      True  04 24, 2015   AR3OHHHW01A8E  B00009W3PA   
4       5      True  03 21, 2015  A2CIEGHZ7L1WWR  B00009W3PA   

                       style     reviewerName  \
0  {'Style:': ' Dryer Vent'}    James. Backus   
1       {'Size:': ' 6-Foot'}           kevin.   
2       {'Size:': ' 6-Foot'}        CDBrannom   
3       {'Size:': ' 6-Foot'}  Calvin E Reames   
4       {'Size:': ' 6-Foot'}   albert j. kong   

                                          reviewText        summary  \
0  I like this as a vent as well as something tha...  Great product   
1                                          good item     Five Stars   
2                     Fit my new LG dryer perfectly.     Five Stars   
3                     Good value for electric dryers   Perfect size   
4                  Price and delivery was excellent.     Five Stars   

   unixReviewTime vote image  
0      1377129600  NaN   NaN  
1      1454889600  NaN   NaN  
2      1438732800  NaN   NaN  
3      1429833600  NaN   NaN  
4      1426896000  NaN   NaN

In [7]:
# Dropping columns not required
review_df.drop(columns=['verified','reviewTime','unixReviewTime','vote','image'], inplace=True)

In [8]:
review_df = review_df[["productID","style","reviewerID","reviewerName","reviewText","summary","rating"]]
review_df.head(3)

productID                      style      reviewerID   reviewerName  \
0  B00009W3I4  {'Style:': ' Dryer Vent'}  A34A1UP40713F8  James. Backus   
1  B00009W3PA       {'Size:': ' 6-Foot'}  A1AHW6I678O6F2         kevin.   
2  B00009W3PA       {'Size:': ' 6-Foot'}   A8R48NKTGCJDQ      CDBrannom   

                                          reviewText        summary  rating  
0  I like this as a vent as well as something tha...  Great product       5  
1                                          good item     Five Stars       5  
2                     Fit my new LG dryer perfectly.     Five Stars       5

# Amazon Appliances meta data

In [9]:
meta_df = pd.read_json('../data/meta_Appliances.json', lines=True)

In [10]:
meta_df.head(3)

category  \
0  [Appliances, Refrigerators, Freezers & Ice Mak...   
1  [Appliances, Refrigerators, Freezers & Ice Mak...   
2              [Appliances, Parts &amp; Accessories]   

                                               tech1  \
0   class="a-keyvalue prodDetTable" role="present...   
1   class="a-keyvalue prodDetTable" role="present...   
2                                                      

                                         description fit  \
0                                                 []       
1  [2 X Tupperware Pure & Fresh Unique Covered Co...       
2                                                 []       

                                               title also_buy tech2  \
0  Tupperware Freezer Square Round Container Set ...       []         
1  2 X Tupperware Pure &amp; Fresh Unique Covered...       []         
2                    The Cigar - Moments of Pleasure       []         

            brand                                            feature  \
0      Tupperware  [Each 3-pc. set includes two 7/8-cup/200 mL an...   
1      Tupperware  [2 X Tupperware Pure & Fresh Unique Covered Co...   
2  The Cigar Book                                                 []   

                                                rank  \
0             [>#39,745 in Appliances (See top 100)]   
1              [>#6,118 in Appliances (See top 100)]   
2  [>#1,861,816 in Home &amp; Kitchen (See Top 10...   

                  also_view details     main_cat similar_item  \
0                        []      {}   Appliances                
1              [B004RUGHJW]      {}   Appliances                
2  [B01HCAVSLK, 1632206579]      {}  Amazon Home                

                date    price        asin  \
0  November 19, 2008           7301113188   
1       June 5, 2016    $3.62  7861850250   
2                     $150.26  8792559360   

                                            imageURL  \
0                                                 []   
1  [https://images-na.ssl-images-amazon.com/image...   
2  [https://images-na.ssl-images-amazon.com/image...   

                                     imageURLHighRes  
0                                                 []  
1  [https://images-na.ssl-images-amazon.com/image...  
2  [https://images-na.ssl-images-amazon.com/image...

In [11]:
meta_df.shape

(30445, 19)

In [29]:
meta_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30239 entries, 0 to 30444
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   productID        30239 non-null  object
 1   title            30239 non-null  object
 2   brand            30239 non-null  object
 3   imageURL         30239 non-null  object
 4   imageURLHighRes  30239 non-null  object
dtypes: object(5)
memory usage: 1.4+ MB


<b>There are no null values in any columns, in the meta data.</b>

In [14]:
# Dropping columns not required
meta_df.drop(columns=['feature','rank','date','description','price','also_view','also_buy','fit','details','similar_item','tech1'], inplace=True)

In [15]:
meta_df = meta_df[["asin","title","brand","imageURL","imageURLHighRes"]]

In [16]:
# Renaming column names : asin to productID.
meta_df.rename(columns={"asin":"productID"},inplace=True)

In [18]:
meta_df.tail()

productID                       title       brand  \
30440  B01HJH92JQ       Bosch 00175338 Switch       Bosch   
30441  B01HJH2I5G  Bosch 00478807 Panel-Facia       Bosch   
30442  B01HJH0J1G    Bosch 00649288 Ice Maker       Bosch   
30443  B01HJHHEA0   Frigidaire 316543810 Knob  Frigidaire   
30444  B01HJH397M   Bosch 00674704 Pump-Drain       Bosch   

                                                imageURL  \
30440                                                 []   
30441  [https://images-na.ssl-images-amazon.com/image...   
30442                                                 []   
30443  [https://images-na.ssl-images-amazon.com/image...   
30444  [https://images-na.ssl-images-amazon.com/image...   

                                         imageURLHighRes  
30440                                                 []  
30441  [https://images-na.ssl-images-amazon.com/image...  
30442                                                 []  
30443  [https://images-na.ssl-images-amazon.com/image...  
30444  [https://images-na.ssl-images-amazon.com/image...

In [19]:
meta_df.productID.nunique()

30239

In [23]:
duplicate_meta = meta_df[meta_df.duplicated(subset="productID" ,keep='first')]
duplicate_meta

productID                                              title  \
210  B00002N5EL  EATON Wiring 39CH-SP-L Arrow Hart 1-Gang Chrom...   
211  B00002N7HY  Leviton 5050 B01-0-000 Electrical Receptacle, ...   
212  B00002N7IL  Leviton 5206 50 Amp, 125/250 Volt, NEMA 10-50R...   
213  B00002N9OE  Amerock BP5322-BJ 1-1/2-Inch Diameter Ceramic ...   
214  B00002NARC       Leviton 5207 125/250V Flush Mount Receptacle   
..          ...                                                ...   
411  B0002Z1FJ8  Electric Oven Knob Kit by Ez-Flo International...   
412  B0002Z1GR4                             Range Hood Motor Blade   
413  B0002Z1GJ2  Whirlpool 326795 Range Broil Element Genuine O...   
414  B00032ATU6               GE 51147 Wireless Remote Transmitter   
415  B00042JSRK                    Engel MT60 AC/DC Fridge-Freezer   

                           brand  \
210                 EATON Wiring   
211                      Leviton   
212                      Leviton   
213                      Amerock   
214                      Leviton   
..                           ...   
411                           GE   
412  Universal Metals Industries   
413                    Whirlpool   
414                           GE   
415                Engel Coolers   

                                              imageURL  \
210                                                 []   
211                                                 []   
212                                                 []   
213                                                 []   
214                                                 []   
..                                                 ...   
411                                                 []   
412                                                 []   
413                                                 []   
414                                                 []   
415  [https://images-na.ssl-images-amazon.com/image...   

                                       imageURLHighRes  
210                                                 []  
211                                                 []  
212                                                 []  
213                                                 []  
214                                                 []  
..                                                 ...  
411                                                 []  
412                                                 []  
413                                                 []  
414                                                 []  
415  [https://images-na.ssl-images-amazon.com/image...  

[206 rows x 5 columns]

In [24]:
meta_df[meta_df['productID']=='B00002N5EL']

productID                                              title  \
4    B00002N5EL  EATON Wiring 39CH-SP-L Arrow Hart 1-Gang Chrom...   
210  B00002N5EL  EATON Wiring 39CH-SP-L Arrow Hart 1-Gang Chrom...   

            brand imageURL imageURLHighRes  
4    EATON Wiring       []              []  
210  EATON Wiring       []              []

<b>There are 206 duplicated values of product ID and there are stacked from index 210 to 415. We can remove these duplicates.</b>

In [25]:
meta_df.drop(duplicate_meta.index, inplace=True)

In [31]:
meta_df[meta_df.duplicated(subset="productID" ,keep='first')]

Empty DataFrame
Columns: [productID, title, brand, imageURL, imageURLHighRes]
Index: []

# EDA